File: android/app/src/main/java/com/personaldiary/android/data/NativeTodoStore.kt
package com.personaldiary.android.data

import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID

/**
 * Local-first native todos (Vault schema v1).
 * File: todos/todos.json → { updated_at, items: [...] }
 * Legacy bare array is migrated on read.
 */
class NativeTodoStore(dataRoot: File) {
    private val file = File(dataRoot, "todos/todos.json")

    fun list(): List<NativeTodo> =
        readDoc().items.sortedWith(compareBy<NativeTodo> { it.done }.thenByDescending { it.updatedAt })

    fun documentUpdatedAt(): String = readDoc().updatedAt

    fun add(text: String): NativeTodo {
        val now = DiaryDates.nowIso()
        val todo = NativeTodo(
            id = UUID.randomUUID().toString(),
            text = text.trim(),
            createdAt = now,
            updatedAt = now,
        )
        val doc = readDoc()
        writeDoc(TodoDocument(updatedAt = now, items = doc.items + todo))
        return todo
    }

    fun upsert(todo: NativeTodo): NativeTodo {
        val now = DiaryDates.nowIso()
        val saved = todo.copy(updatedAt = now)
        val doc = readDoc()
        val items = doc.items.toMutableList()
        val idx = items.indexOfFirst { it.id == saved.id }
        if (idx >= 0) items[idx] = saved else items += saved
        writeDoc(TodoDocument(updatedAt = now, items = items))
        return saved
    }

    fun setDone(id: String, done: Boolean): NativeTodo? {
        val doc = readDoc()
        val idx = doc.items.indexOfFirst { it.id == id }
        if (idx < 0) return null
        val updated = doc.items[idx].copy(done = done, updatedAt = DiaryDates.nowIso())
        val items = doc.items.toMutableList().also { it[idx] = updated }
        writeDoc(TodoDocument(updatedAt = updated.updatedAt, items = items))
        return updated
    }

    fun delete(id: String) {
        val doc = readDoc()
        writeDoc(TodoDocument(updatedAt = DiaryDates.nowIso(), items = doc.items.filterNot { it.id == id }))
    }

    fun replaceAll(items: List<NativeTodo>) {
        writeDoc(
            TodoDocument(
                updatedAt = items.maxOfOrNull { it.updatedAt }.orEmpty().ifBlank { DiaryDates.nowIso() },
                items = items,
            ),
        )
    }

    fun readRaw(): String {
        val doc = readDoc()
        if (!file.isFile || file.readText(Charsets.UTF_8).trimStart().startsWith("[")) {
            writeDoc(doc)
        }
        return if (file.isFile) file.readText(Charsets.UTF_8) else encodeDoc(TodoDocument("", emptyList()))
    }

    fun writeRaw(json: String) {
        writeDoc(parseDoc(json))
    }

    /** Apply remote todos payload (`json` field) with server document timestamp. */
    fun applyRemote(jsonField: String, documentUpdatedAt: String) {
        val parsed = parseDoc(jsonField)
        writeDoc(
            TodoDocument(
                updatedAt = documentUpdatedAt.ifBlank { parsed.updatedAt },
                items = parsed.items,
            ),
        )
    }

    private data class TodoDocument(
        val updatedAt: String,
        val items: List<NativeTodo>,
    )

    private fun readDoc(): TodoDocument {
        if (!file.isFile) return TodoDocument(updatedAt = "", items = emptyList())
        return try {
            parseDoc(file.readText(Charsets.UTF_8))
        } catch (_: Exception) {
            TodoDocument(updatedAt = "", items = emptyList())
        }
    }

    private fun writeDoc(doc: TodoDocument) {
        file.parentFile?.mkdirs()
        file.writeText(encodeDoc(doc), Charsets.UTF_8)
    }

    companion object {
        private fun parseDoc(raw: String): TodoDocument {
            val trimmed = raw.trim()
            if (trimmed.isEmpty()) return TodoDocument("", emptyList())
            val root = when {
                trimmed.startsWith("[") ->
                    JSONObject()
                        .put("updated_at", "")
                        .put("items", JSONArray(trimmed))
                else -> JSONObject(trimmed)
            }
            if (root.has("json") && !root.has("items")) {
                val inner = root.opt("json")
                val innerObj = when (inner) {
                    is JSONObject -> inner
                    is String -> {
                        val s = inner.trim()
                        when {
                            s.startsWith("[") ->
                                JSONObject()
                                    .put("updated_at", root.optString("updated_at"))
                                    .put("items", JSONArray(s))
                            s.startsWith("{") -> JSONObject(s)
                            else -> JSONObject().put("items", JSONArray())
                        }
                    }
                    else -> JSONObject().put("items", JSONArray())
                }
                if (innerObj.has("items")) {
                    val items = parseItems(innerObj.optJSONArray("items") ?: JSONArray())
                    val updated = root.optString("updated_at")
                        .ifBlank { innerObj.optString("updated_at") }
                        .ifBlank { items.maxOfOrNull { it.updatedAt }.orEmpty() }
                    return TodoDocument(updated, items)
                }
            }
            val items = parseItems(root.optJSONArray("items") ?: JSONArray())
            val updated = root.optString("updated_at").ifBlank {
                items.maxOfOrNull { it.updatedAt }.orEmpty()
            }
            return TodoDocument(updated, items)
        }

        private fun parseItems(arr: JSONArray): List<NativeTodo> =
            buildList {
                for (i in 0 until arr.length()) {
                    val o = arr.optJSONObject(i) ?: continue
                    val due = o.opt("due_at")
                    val dueAt = when {
                        due == null || due == JSONObject.NULL -> ""
                        else -> due.toString().trim().let { if (it == "null") "" else it }
                    }
                    add(
                        NativeTodo(
                            id = o.optString("id").ifBlank { UUID.randomUUID().toString() },
                            text = o.optString("text"),
                            detail = o.optString("detail"),
                            done = o.optBoolean("done", false),
                            kind = o.optString("kind", "task").ifBlank { "task" },
                            dueAt = dueAt,
                            priority = o.optInt("priority", 0).coerceIn(0, 3),
                            urgency = o.optInt("urgency", 0).coerceIn(0, 3),
                            createdAt = o.optString("created_at"),
                            updatedAt = o.optString("updated_at"),
                        ),
                    )
                }
            }
    }

    private fun encodeDoc(doc: TodoDocument): String {
        val arr = JSONArray()
        for (t in doc.items) {
            val o = JSONObject()
                .put("id", t.id)
                .put("text", t.text)
                .put("detail", t.detail)
                .put("done", t.done)
                .put("kind", t.kind.ifBlank { "task" })
                .put("priority", t.priority.coerceIn(0, 3))
                .put("urgency", t.urgency.coerceIn(0, 3))
                .put("created_at", t.createdAt)
                .put("updated_at", t.updatedAt)
            if (t.dueAt.isBlank()) o.put("due_at", JSONObject.NULL) else o.put("due_at", t.dueAt)
            arr.put(o)
        }
        return JSONObject()
            .put("updated_at", doc.updatedAt)
            .put("items", arr)
            .toString(2)
    }

    private fun parseDoc(raw: String): TodoDocument = Companion.parseDoc(raw)
}

File: src/storage/native_todo_store.py
"""Native todos — vault todos/todos.json (aligned with Android NativeTodoStore)."""
from __future__ import annotations

import json
import uuid
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Optional


def _now() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


@dataclass
class NativeTodo:
    id: str
    text: str
    detail: str = ""
    done: bool = False
    kind: str = "task"  # task | note | errand | other
    due_at: str = ""
    priority: int = 0
    urgency: int = 0
    created_at: str = ""
    updated_at: str = ""


@dataclass
class TodoDocument:
    updated_at: str = ""
    items: list[NativeTodo] = field(default_factory=list)


class NativeTodoStore:
    def __init__(self, data_root: Path) -> None:
        self.file = Path(data_root) / "todos" / "todos.json"

    def list(self) -> list[NativeTodo]:
        items = sorted(self._read().items, key=lambda t: t.updated_at, reverse=True)
        return sorted(items, key=lambda t: t.done)

    def add(self, text: str) -> NativeTodo:
        now = _now()
        todo = NativeTodo(
            id=str(uuid.uuid4()),
            text=text.strip(),
            created_at=now,
            updated_at=now,
        )
        doc = self._read()
        self._write(TodoDocument(updated_at=now, items=doc.items + [todo]))
        return todo

    def upsert(self, todo: NativeTodo) -> NativeTodo:
        now = _now()
        saved = NativeTodo(**{**asdict(todo), "updated_at": now})
        doc = self._read()
        items = list(doc.items)
        idx = next((i for i, t in enumerate(items) if t.id == saved.id), -1)
        if idx >= 0:
            items[idx] = saved
        else:
            items.append(saved)
        self._write(TodoDocument(updated_at=now, items=items))
        return saved

    def set_done(self, todo_id: str, done: bool) -> Optional[NativeTodo]:
        doc = self._read()
        items = list(doc.items)
        idx = next((i for i, t in enumerate(items) if t.id == todo_id), -1)
        if idx < 0:
            return None
        now = _now()
        updated = NativeTodo(**{**asdict(items[idx]), "done": done, "updated_at": now})
        items[idx] = updated
        self._write(TodoDocument(updated_at=now, items=items))
        return updated

    def delete(self, todo_id: str) -> None:
        doc = self._read()
        now = _now()
        self._write(
            TodoDocument(
                updated_at=now,
                items=[t for t in doc.items if t.id != todo_id],
            )
        )

    def _read(self) -> TodoDocument:
        if not self.file.is_file():
            return TodoDocument()
        try:
            return self._parse(self.file.read_text(encoding="utf-8"))
        except Exception:  # noqa: BLE001
            return TodoDocument()

    def _write(self, doc: TodoDocument) -> None:
        self.file.parent.mkdir(parents=True, exist_ok=True)
        self.file.write_text(self._encode(doc), encoding="utf-8")

    @staticmethod
    def _parse(raw: str) -> TodoDocument:
        trimmed = raw.strip()
        if not trimmed:
            return TodoDocument()
        data: Any = json.loads(trimmed)
        if isinstance(data, list):
            items_raw = data
            updated_at = ""
        else:
            items_raw = data.get("items") or []
            updated_at = str(data.get("updated_at") or "")
        items: list[NativeTodo] = []
        for obj in items_raw:
            if not isinstance(obj, dict):
                continue
            items.append(
                NativeTodo(
                    id=str(obj.get("id") or uuid.uuid4()),
                    text=str(obj.get("text") or ""),
                    detail=str(obj.get("detail") or ""),
                    done=bool(obj.get("done")),
                    kind=str(obj.get("kind") or "task"),
                    due_at=str(obj.get("due_at") or "") if obj.get("due_at") is not None else "",
                    priority=int(obj.get("priority") or 0),
                    urgency=int(obj.get("urgency") or 0),
                    created_at=str(obj.get("created_at") or ""),
                    updated_at=str(obj.get("updated_at") or ""),
                )
            )
        return TodoDocument(updated_at=updated_at, items=items)

    @staticmethod
    def _encode(doc: TodoDocument) -> str:
        payload = {
            "updated_at": doc.updated_at,
            "items": [
                {
                    "id": t.id,
                    "text": t.text,
                    "detail": t.detail,
                    "done": t.done,
                    "kind": t.kind,
                    "due_at": t.due_at or None,
                    "priority": t.priority,
                    "urgency": t.urgency,
                    "created_at": t.created_at,
                    "updated_at": t.updated_at,
                }
                for t in doc.items
            ],
        }
        return json.dumps(payload, ensure_ascii=False, indent=2)
